src/test/scala/zio/nio/ChannelSuite.scala

```diff
@@ -1,54 +1,130 @@
-import zio.{ DefaultRuntime, IO }
+import zio._
 import testz.{ Harness, assert }
 
 object ChannelSuite extends DefaultRuntime {
 
   def tests[T](harness: Harness[T]): T = {
     import harness._
-    section(test("read/write") { () =>
-      val inetAddress = InetAddress.localHost
-        .flatMap(iAddr => SocketAddress.inetSocketAddress(iAddr, 13370))
-
-      def echoServer: IO[Exception, Unit] =
-        for {
-          address <- inetAddress
-          sink    <- Buffer.byte(3)
-          server  <- AsynchronousServerSocketChannel()
-          _       <- server.bind(address)
-          worker  <- server.accept
-          _       <- worker.readBuffer(sink)
-          _       <- sink.flip
-          _       <- worker.writeBuffer(sink)
-          _       <- worker.close
-          _       <- server.close
-        } yield ()
-
-      def echoClient: IO[Exception, Boolean] =
-        for {
-          address  <- inetAddress
-          src      <- Buffer.byte(3)
-          client   <- AsynchronousSocketChannel()
-          _        <- client.connect(address)
-          sent     <- src.array
-          _        = sent.update(0, 1)
-          _        <- client.writeBuffer(src)
-          _        <- src.flip
-          _        <- client.readBuffer(src)
-          received <- src.array
-          _        <- client.close
-        } yield sent.sameElements(received)
-
-      val testProgram: IO[Exception, Boolean] = for {
-        serverFiber <- echoServer.fork
-        clientFiber <- echoClient.fork
-        _           <- serverFiber.join
-        same        <- clientFiber.join
-      } yield same
-
-      assert(unsafeRun(testProgram))
-    })
+    section(
+      test("read/write") { () =>
+        val inetAddress = InetAddress.localHost
+          .flatMap(iAddr => SocketAddress.inetSocketAddress(iAddr, 13370))
+
+        def echoServer: IO[Exception, Unit] =
+          for {
+            address <- inetAddress
+            sink    <- Buffer.byte(3)
+            server  <- AsynchronousServerSocketChannel()
+            _       <- server.bind(address)
+            _ <- server.accept
+                  .bracket(_.close.ignore *> server.close.ignore) { worker =>
+                    worker.readBuffer(sink) *>
+                      sink.flip *>
+                      worker.writeBuffer(sink)
+                  }
+                  .fork
+          } yield ()
+
+        def echoClient: IO[Exception, Boolean] =
+          for {
+            address  <- inetAddress
+            src      <- Buffer.byte(3)
+            client   <- AsynchronousSocketChannel()
+            _        <- client.connect(address)
+            sent     <- src.array
+            _        = sent.update(0, 1)
+            _        <- client.writeBuffer(src)
+            _        <- src.flip
+            _        <- client.readBuffer(src)
+            received <- src.array
+            _        <- client.close
```

Comment:
Should client also be bracketed?

Reply:
for this case it doesn't need to be but I can do it